test/hotspot/gtest/nmt/test_vmatree.cpp

```diff
@@ -0,0 +1,516 @@
+  NativeCallStackStorage::StackIndex si2 = ncs.push(stack2);
+
+  // Adjacent reservations should result in exactly 2 nodes
+  auto adjacent_2_nodes = [&](const VMATree::RegionData& rd) {
+    Tree tree;
+    for (int i = 0; i < 100; i++) {
+      tree.reserve_mapping(i * 100, 100, rd);
+    }
+    int found_nodes = 0;
+    treap(tree).visit_range_in_order(0, 999999, [&](Node* x) {
+      found_nodes++;
+    });
+    EXPECT_EQ(2, found_nodes) << "Adjacent reservations should result in exactly 2 nodes";
+
+    // Reserving the exact same space again should result in still having only 2 nodes
+    for (int i = 0; i < 100; i++) {
+      tree.reserve_mapping(i * 100, 100, rd);
+    }
+    found_nodes = 0;
+    treap(tree).visit_range_in_order(0, 999999, [&](Node* x) {
+      found_nodes++;
+    });
+    EXPECT_EQ(2, found_nodes) << "Adjacent reservations should result in exactly 2 nodes";
+
+    // Do it backwards instead.
+    Tree tree2;
+    for (int i = 99; i >= 0; i--) {
+      tree2.reserve_mapping(i * 100, 100, rd);
+    }
+    found_nodes = 0;
+    treap(tree2).visit_range_in_order(0, 999999, [&](Node* x) {
+      found_nodes++;
+    });
+    EXPECT_EQ(2, found_nodes) << "Adjacent reservations should result in exactly 2 nodes";
+  };
+
+  { // Overlapping reservations should also only result in 2 nodes.
+    VMATree::RegionData rd{si1, mtTest};
+    Tree tree2;
+    for (int i = 99; i >= 0; i--) {
+      tree2.reserve_mapping(i * 100, 101, rd);
+    }
+    int found_nodes = 0;
+    treap(tree2).visit_range_in_order(0, 999999, [&](Node* x) {
+      found_nodes++;
+    });
+    EXPECT_EQ(2, found_nodes) << "Adjacent reservations should result in exactly 2 nodes";
+  }
+
+  // After removing all ranges we should be left with an entirely empty tree
+  auto remove_all_leaves_empty_tree = [&](const VMATree::RegionData& rd) {
+    Tree tree;
+    tree.reserve_mapping(0, 100*100, rd);
+    for (int i = 0; i < 100; i++) {
+      tree.release_mapping(i*100, 100);
+    }
+    EXPECT_EQ(nullptr, treap_root(tree)) << "Releasing all memory should result in an empty tree";
+
+    // Other way around
+    tree.reserve_mapping(0, 100*100, rd);
+    for (int i = 99; i >= 0; i--) {
+      tree.release_mapping(i*100, 100);
+    }
+    EXPECT_EQ(nullptr, treap_root(tree)) << "Releasing all memory should result in an empty tree";
+  };
+
+  // Committing in middle of reservation ends with a sequence of 4 nodes
+  auto commit_middle = [&](const VMATree::RegionData& rd) {
+    Tree tree;
+    tree.reserve_mapping(0, 100, rd);
+    tree.commit_mapping(50, 25, rd);
+
+    size_t found[16];
+    size_t wanted[4] = {0, 50, 75, 100};
+    auto exists = [&](size_t x) {
+      for (int i = 0; i < 4; i++) {
+        if (wanted[i] == x) return true;
+      }
+      return false;
+    };
+    int i = 0;
+    treap(tree).visit_range_in_order(0, 300, [&](Node* x) {
+      if (i < 16) {
+        found[i] = x->key();
+      }
+      i++;
+    });
```

Comment:
This doesn't really test the state, nor the stack. It also seems to be a lot of code for a single-use test. Similar in other tests.

---

Proposal: since you have the recurring pattern of doing something with the tree, then checking its expected state, write a check function in the fixture (e.g. assert_tree_state() ) that does that.

Then use it like this:
```
// Committing in middle of reservation ends with a sequence of 4 nodes
TEST_VM_F(VMATreeTest, commit_in_middle_of_reservation) {

	... reserve, commit in middle, then

	const AddressState expected_states[] = {
		{ 0,  .... },
		{ 25,  .... },
		{ 50,  .... },
		{ 100,  .... } };
	assert_tree_state(expected_states, 4);
}
```

And to preserve your sanity when constructing AddressState with its many components, you can add a helper that allows that in a human-friendly form.

For example:
A string that encodes flag, stack, and reserved/Commit state in three letters. 
First letter: A-H, let this be one of 8 selected MEMFLAGS or - for mtNone
Second letter: a-d, let this be one of 4 selected stacks, or - for no stack/empty
Third letter: R or C , resreved, committed or none, or - for unreserved
Season to taste.

Then, e.g. for reserving 0..100, and committing 25..50, you write:

```
const AddressState expected_states[] = {
	makestate( 0, "---", "AaR"),
	makestate(25, "AaR", "AaC"),
	makestate(50, "AaC", "AaR"),
	makestate( 0, "AaR", "---") };
assert_tree_state(expected_states, 4);
```

Now you can write a bunch of corner-case tests without being too verbose, intent is immediately clear. As an added bonus, it checks for tree state *exactly*, e.g. which nodes are part of the tree, which are not, in which order, etc.

If you reuse that string format for reserving etc, you could do this:

```
    reserve_or_commit(0, 100, "AaR");
    reserve_or_commit(25, 50, "AaC");
```

or just 

```
    do(0, 100, "AaR");
    do(25, 50, "AaC");
```

From there one, one could even auto-build the expected state, but if too much automatism goes into the test, this increases the possibility for errors creeping in one never finds because they make the test go green.

Reply:
While going for a walk after writing this, I realized an alternative to a string would be just to define those AddressState directly as global constants.

static const AddressState AaR = ....

With 4 flags, 2 stacks and 3 states, this would come to 24 states. With a macro, this could even be easier.

Up to you.

Reply:
I'm going to take some time to digest these ideas. I'm not a fan of the string-based approach, I much prefer longer but more obvious code. The latter approach might work out. Still, a general tool is to me not preferable. The typical case is probably not that you need to read every test case, but a specific one which fails. Having some repetition in other tests shouldn't bother you then, but having to jump into a generalized DSL for state assertions do.

>This doesn't really test the state, nor the stack. It also seems to be a lot of code for a single-use test. Similar in other tests.

Sorry, I don't understand what you mean by this.

Reply:
>> This doesn't really test the state, nor the stack. It also seems to be a lot of code for a single-use test. Similar in other tests.

> Sorry, I don't understand what you mean by this.


You test that after committing, we have four nodes. You don't test that we have reserved-committed-reserved, nor the associated stacks (all equal), nor the flags. The test would be more interesting if you were to verify the region data, too.

Reply:
> I'm going to take some time to digest these ideas. I'm not a fan of the string-based approach, I much prefer longer but more obvious code. The latter approach might work out. Still, a general tool is to me not preferable. The typical case is probably not that you need to read every test case, but a specific one which fails. Having some repetition in other tests shouldn't bother you then, but having to jump into a generalized DSL for state assertions do.

But is this really so far away from what you are doing, just a lot less code?

Thinking about it, 24 different region states is probably too much. If we go for, say, 12 states, that increases the chance of merging in the random tests. So you pre-define, constant, 12 different region data with clear names. Those you use to reserve/commit, and to check the tree state.